File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        vector<string> words; // To store individual words
        string currentWord;   // To build the current word
        string result;        // To store the final result

        // Step 1: Parse the string to extract words
        for (char ch : s) {
            if (ch == ' ') {
                if (!currentWord
                         .empty()) { // If a word is formed, add it to the list
                    words.push_back(currentWord);
                    currentWord.clear(); // Reset for the next word
                }
            } else {
                currentWord += ch; // Add non-space characters to current word
            }
        }
        if (!currentWord.empty()) { // Add the last word if present
            words.push_back(currentWord);
        }

        // Step 2: Reverse the list of words and construct the result
        for (int i = words.size() - 1; i >= 0; --i) {
            result += words[i];
            if (i != 0)
                result += " "; // Add a single space between words
        }

        return result;
    }
};

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    public String reverseWords(String s) {
        Stack<String> st = new Stack<>();
        StringBuilder word = new StringBuilder(); // Store current word
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch == ' ') {
                if(word.length()>0) st.push(word.toString()); //check only if word exist
                word.setLength(0); // Clear string word for next word       
                }
                else word.append(ch); 
        }
        if(word.length()>0) st.push(word.toString());  // Push last word
        // Construct the reversed sentence
        StringBuilder ans = new StringBuilder();
        while (!st.isEmpty()) {
            ans.append(st.pop()).append(" ");
        }
        return ans.toString().trim(); // Remove trailing space and return result
    }
}